### PR TITLE
scripts/download.pl: make the download tool configurable

### PR DESCRIPTION
**Pull request by Ansuel**
Introduce a new option in the "Advanced configuration options" to
configure a custom download tool.

By declaring a string in "Use custom download tool" an user can force
what command to use to download package. With the string empty the
default tool used is curl, with wget as a fallback if not available.

download.pl supports 3 tools officially aria2c, curl and wget.
If one of the tool is used in this config, download.pl will use the
default args to make use of them.

If the provided string is different than aria2c, curl or wget, the command
is used as is and the download url will be appended at the end of such command.

While at it also tweak the tool selection logic and chose the tool only
once when the script is called and move aria2c specific variables in the
relevant section.

Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>

----

This was requested by an user on IRC as aria2c can't be used for proxy... For some reason he use a distro that doesn't support removing aria2c so nothing is downloaded.

2 example alternative to use this option.
(an user can provide the single program and download.pl will use default args if supported or just take what is passed in this string)

image
image

### CODE REVIEW

**Comment by Ansuel** [comment]
@mpratt14 since you love this kind of thing random quote if you have some logic hint :D

**Comment by neheb** [comment]
I'll just mention that only wget is in prereq-build.mk for some reason.

**Comment by ynezz** [comment]
> By default this option is set to aria2c
> This was requested by an user on IRC as aria2c can't be used for proxy

It looks like an oxymoron to me, how can you use a clearly feature incomplete tool by default? IMO this aria2c looks like an [experimental feature](https://github.com/openwrt/openwrt/blob/master/config/Config-build.in#L6) and should be probably treated as such?

File: config/Config-devel.in [hunk 1]
```diff
@@ -17,6 +17,30 @@ menuconfig DEVEL
 		  Store built firmware images and filesystem images in this directory.
 		  If not set, uses './bin/$(BOARD)'
 
+	choice
+		prompt "Preferred Download tool" if DEVEL
+		default DOWNLOAD_TOOL_PREFER_ARIA2C
+		help
+		  Set the preferred tool to use to download package on the buildroot.
+		  If the tool is not present in the system, the system will fallback
+		  to the others until it find one.
```
Comment (ynezz) [inline config/Config-devel.in:26]: Do we really need to clutter the build system with such logic? That list might get miles long soon, so what about simply admitting, that we support sane defaults, `wget` and `curl` out of the box and treat any other tool as a custom downloader? So if user wants to use tool `foo`, then simply just use that. If I explicitly enable tool `foo`, then simply error out if it's not available, no need for fallback handling, keep it simple.

**Comment by ynezz** [comment]
> IMO this aria2c looks like an [experimental feature](https://github.com/openwrt/openwrt/blob/master/config/Config-build.in#L6) and should be probably treated as such?

Scratch that, looking at the amount of changes needed to support aria2c and fallout it causes, I would simply handle it in a more generic way. I would add a config option which would allow setting of a custom download tool and simply use that, no matter what that tool is, no need for handling any fallbacks as well.

**Comment by mpratt14** [comment]
> I'll just mention that only wget is in prereq-build.mk for some reason.

i have a commit in my huge RFC that handles this

File: config/Config-devel.in (hunk 1 above)
Comment (mpratt14) [reply config/Config-devel.in:26]: I agree that curl should return as the default, but some people need aria

I don't think that there even exists that many tools for just doing a download so the list would not get much larger. However, I didn't even know that aria existed so...

**Comment by mpratt14** [comment]
the Perl stuff is very fancy, but I don't think it's necessary, and its a little complex

we should be able to handle that part with Make

consider the situation with prereq-build.mk again:

 - ifeq statement for each config option to check for each tool
 - make staging_dir/host/.prereq-build depend on .config

**Comment by Ansuel** [comment]
@ynezz yes honestly i also had some concern about using a fallback... but at the end i decided to propose this solution... 

But considering the final change, think I will just "keep it simple" TM, and force the tool and fail if not available.

@mpratt14 don't know if we should revert to curl by default, probably yes. (with this feature the chinese problem that require aria2c can still use it while revert to a more and simple solution for everyone else... but honestly i notice much quicker download with aria2c, but also a drawback since it will put more pressure on our server due to the multiple connection thing)

Also the prereq thing doesn't apply, the logic is to find whatever is available, this is why we have only wget in prereq, it's to at least have ONE tool to download stuff.

**Comment by ynezz** [comment]
> don't know if we should revert to curl by default

IIUC then it's, prefer `curl` in case it's available and use `wget` as mandatory default, that's why it's being checked by prereq.

**Comment by Ansuel** [comment]
> > don't know if we should revert to curl by default
> 
> IIUC then it's, prefer `curl` in case it's available and use `wget` as mandatory default, that's why it's being checked by prereq.

correct with aria2c support this has changed to 
aria2c -> curl -> wget

I would revert this to make curl the preferred one, wget as fallback and aria2c as a selectable option in menu config. Think it's the best solution for everyone.

**Comment by mpratt14** [comment]
I still think prereq-build is a great way to handle it, maybe I didnt explain well enough

optionally check for aria first, then curl, then wget, first successful one gets a symlink installed in staging_dir/host/bin, then download.pl simply looks for the symlinks in order, except look for aria symlink first when that config is selected

File: config/Config-devel.in (hunk 1 above)
Comment (aparcar) [reply config/Config-devel.in:26]: I missed the conversation, why do some people need aria?

Comment (Ansuel) [reply config/Config-devel.in:26]: Mainly Chinese users that have slow download connection and mirrors problem. Aria2c can check every mirror and select a good one and increase download speed.

**Comment by Ansuel** [comment]
@mpratt14 seems a nice idea can you do the tweak after this change? I also don't like the fact that we check the tool presence using perl... i agree that makefile should do this.

**Comment by Ansuel** [comment]
@mpratt14 anyway was experimenting with your idea and is problematic... the idea is to add optional prereq right? How that would work when aria2c is not installed in the system and then an user installs it? prereq won't be refreshed sooo i think the correct solution is to check them at runtime... one thing to improve would be make the check one time instead of for each package... but i don't think it's worth it as you don't call download.pl that often...

**Comment by Ansuel** [comment]
@ynezz @neheb can you check this variant? Think it's better and should make everyone happy... (hope chinese user will notice this pr and read the instruction on how to use aria2c)

**Comment by mpratt14** [comment]
> when aria2c is not installed in the system and then a user installs it

prereq set based on config and depending on the config, not whether the program exists (updating TestHostCommand to allow multiple alternatives would help too)

you're creating config option `DOWNLOAD_TOOL_PREFER_ARIA2C` so you can use that to decide whether there is a prereq for aria

in my draft PR I have in prereq-build.mk

```
ifneq ($(CONFIG_BUILD_HOST_PYTHON),y)
   (require python)
endif
```

**Comment by Ansuel** [comment]
@mpratt14 oh ok so prereq will update on config change and actually warn that the tool is not installed... interesting idea

**Comment by mpratt14** [comment]
> can you do the tweak after this change?

and yeah it can be done later of course

**Comment by Ansuel** [comment]
@neheb @ynezz is it ok? I would merge this since it looks a way better implementation than it's currently on

**Comment by neheb** [comment]
Seems fine to me. I'd like @jow- ''s input.

File: config/Config-devel.in [hunk 2]
```diff
@@ -17,6 +17,41 @@ menuconfig DEVEL
 		  Store built firmware images and filesystem images in this directory.
 		  If not set, uses './bin/$(BOARD)'
 
+	menuconfig DOWNLOAD_TOOL_USE_CUSTOM
+		bool
```
Comment (ynezz) [inline config/Config-devel.in:21]: I would name it `DOWNLOAD_TOOL_CUSTOM` and make it string, so keeping it simple and future proof. That way it doesn't matter if you use aria2c, super-curl3, spider-fu8 or monkey-get3000.

Comment (Ansuel) [reply config/Config-devel.in:21]: sooo dropping the selection and just make the user insert a string? download.pl still needs to have support for it, that is why i wanted a selection so the user know what is supported.

Comment (ynezz) [reply config/Config-devel.in:21]: > download.pl still needs to have support for it

Comment (Ansuel) [reply config/Config-devel.in]: @ynezz ok tried to do my best to find a ""future proof" and keep support with the 3 tools we support. What do you think?

**Comment by Ansuel** [comment]
@ynezz @mpratt14 can you guys check this? For me it's ready but wonder if you have other comments about my implementation

File: scripts/download.pl [hunk 3]
```diff
@@ -85,16 +87,42 @@ sub tool_present {
 	return $present
 }
 
+sub select_tool {
+	$custom_tool =~ tr/"//d;
+	if ($custom_tool) {
+		return $custom_tool;
+	}
```
Comment (mpratt14) [inline scripts/download.pl:94]: just a question, the config is a string with only `""` when it's empty, so in perl it's not a null string? and removing the characters makes it a null string?

anyway, I think this regex should be done when custom_tool is first defined, if it even needs to be done at all. I don't know Perl very well but it apparently ignores double quotes for the most part, compared with a shell where double quotes decides how much it is expanded

Comment (Ansuel) [reply scripts/download.pl:94]: well lost too much times discovering that... ENV variable are with "" so i have to strip them to be detected as NULL ahahha

Comment (mpratt14) [reply scripts/download.pl:94]: not all variables in ENV, but just the configs that are strings right?

**Comment by Ansuel** [comment]
Problem is that you can't use the same command for all of the tools so you
need special handling for them anyway.

Il Sab 1 Ott 2022, 17:11 Michael Pratt ***@***.***> ha
scritto:

File: scripts/download.pl (hunk 3 above)
Comment (Ansuel) [reply scripts/download.pl:94]: yes as they will be set but set to `""` as empty

**Comment by Ansuel** [comment]
@neheb any comments on the new implementation?

**Comment by Ansuel** [comment]
@ynezz ping? would love a final review on this

**Pull request merged**